README.md
# Simple HTTP Deploy tool
It will update package version in docker-compose and run commands:
```bash
docker-compose -f custom.docker-compose.yml up -d
```
Previous docker compose file will be saved as `{previousDockerComposePath}_{previousImageVersion}`

## Configuration
Configuration by environment variables:
- **DEPLOYER_PORT** - port that Express listen
- **DEPLOYER_CONFIG_PATH** - path to JSON file with projects config

JSON config file should contain array, each element have to implement [ConfigInterface](src/ConfigInterface.ts)

## Requesting

#### Request
POST / 

Content-Type: application/json

```json
{
  "name": "image-name",
  "tag": "docker-registry.private.com/image-name:1.0.0"
}
```

#### Response
Content-Type: application/json

| Http Code 	| JSON Code 	| Message                                        	| What to do?                                                                                                                                                      	|
|-----------	|-----------	|------------------------------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------------------	|
| 200       	| 0         	| Successful                                     	| Drink beer                                                                                                                                                       	|
| 403       	| -1        	| Forbidden                                      	| Try to pass valid                                                                                                                                                	|
| 422       	| 101       	| Missing name or tag in body                    	| Put JSON body with name and tag fields                                                                                                                           	|
| 422       	| 102       	| {name} deploy does not exist                   	| Configure {name} deploy on server                                                                                                                                	|
| 501       	| 201       	| Compose file does not exist                    	| Fix path to docker-compose file in global configuration                                                                                                          	|
| 501       	| 202       	| Can not find image name in docker-compose file 	| Make sure that image specified in docker-compose file. Image name should be placed between `"` and have 3-digit version after `:` (example: `"image-name:1.0.0`) 	|
| 501       	| 203       	| Invalid {type} time: {invalidTimeFormat}       	| Make sure that time rules in configuration if specified in format `HH:MM:SS`                                                                                     	|
| 500       	| 301       	| BeforeDeploy Error                             	| Error in some BeforeDeploy script. See deploy logs for details and improve script                                                                                	|
| 500       	| 302       	| DockerDeploy Error                             	| Error while executing `docker-compose up -d`. See deploy logs for details                                                                                        	|
| 500       	| 303       	| AfterDeploy Error                              	| Error in some AfterDeploy script. See deploy logs for details and improve script                                                                                 	|
| 503       	| 403       	| Update not available                           	| Try run update considering time specified in time rules                                                                                        	                  |

**Notice: message can be changed in minor version. You should not parse it and depend your CI process on it.**

Example:
```json
{
  "code": 0,
  "message": "Successful"
}
```

#### Request
GET /status/:projectName

#### Response
Content-Type: application/json

| Http Code | JSON Code   | Message                     | What to do?                                             |
|---------- |------------ |---------------------------  |-------------------------------------------------------- |
| 200       | 0           | Successful                  | Order Pizza                                             |
| 501       | 201       	| Compose file does not exist | Fix path to docker-compose file in global configuration |
| 503       | any number  | Docker Error                | Check executing docker-compose ps on server             |    

**Notice: message can be changed in minor version. You should not parse it and depend your CI process on it.**

Example:
```json
{
  "code": 0,
  "message": "Successful",
  "containers": [
    {
      "name": "app-1",
      "state": "Up"
    },
    {
      "name": "app-2",
      "state": "Restarting"
    }
  ]
}
```

## Author
[Alexander <horat1us> Letnikow](mailto:reclamme@gmail.com)

## License
MIT


CHANGELOG.md
# Change Log

## 1.9.0
- Refactor `monitoring` route using [express-monitoring](https://www.npmjs.com/package/express-monitoring)
package to monitor application configuration and docker engine status.

## 1.2.0
- Add `beforeDeploy` and `afterDeploy` sections to project configuration. It can be null or string[] (array).
- Add `port` section to global configuration (loaded from JSON). 
If no port provided in JSON environment variable **DEPLOYER_PORT** will be used instead.
- Add detailed response types and unique code for each error type (see details in [README](./README.md))


src/routes/upgrade.ts
import * as helpers from "../helpers";
import { ConfigEntity, ConfigInterface } from "../data/ConfigInterface";
import { UpdateInterface } from "../data/UpdateInterface";
import { Request, Response } from "express";

const fs = require("fs")

export const upgrade = (globalConfig: ConfigInterface) => async (request: Request, response: Response): Promise<Response> => {
    const body: UpdateInterface = request.body;
    if (!body.name || !body.tag) {
        return response.status(422).json({ code: 101, message: "Missing name or tag in body" }).send();
    }

    const config: ConfigEntity | undefined = globalConfig.projects.find((config) => config.name === body.name);
    if (!config) {
        console.warn(`Can not find project with name ${body.name}`);
        return response.status(422).json({ code: 102, message: `${body.name} deploy does not exist` }).send();
    }
    console.log(`Begin deploy ${body.name}`);

    if (!fs.existsSync(config.path)) {
        console.error(`File ${config.path} does not exist`);
        return response.status(501).json({ code: 201, message: "Compose file does not exist" }).send();
    }
    if (config.time) {
        try {
            if (!helpers.checkTime(helpers.getCurrentTime(), config.time)) {
                return response.status(503).json({ code: 403, message: "Update not available" }).send();
            }
        } catch (error) {
            return response.status(501).json({ code: 203, message: error.message }).send();
        }
    }
    const yaml = fs.readFileSync(config.path).toString();
    const previousTagRegExp = new RegExp(`image:\\s+".*${body.name}:(\\d+\\.\\d+\\.\\d+)\"`, "g");
    const match = yaml.match(previousTagRegExp);

    if (!match || !match.length) {
        return response.status(501).json({ code: 202, message: `Can not find image name in docker-compose file` }).send();
    }
    const newYaml = yaml.replace(previousTagRegExp, `image: "${body.tag}"`);
    const previousVersion = match[0].match(/:(\d+\.\d+\.\d+)"/)[1];

    fs.renameSync(config.path, `${config.path}_${previousVersion}`);
    fs.writeFileSync(config.path, newYaml);

    try {
        config.beforeDeploy && console.log("Executing beforeDeploy");
        await helpers.executeScripts(config.beforeDeploy);
    }
    catch (error) {
        console.error(error);
        return response.status(500).json({ code: 301, message: "BeforeDeploy Error" }).send();
    }

    try {
        await helpers.exec(`docker-compose -f ${config.path} up -d`);
    }
    catch (error) {
        console.error(error.stderr ? error.stderr : error);
        return response.status(500).json({ code: 302, message: "DockerDeploy Error" }).send();
    }

    try {
        config.afterDeploy && console.log("Executing afterDeploy");
        await helpers.executeScripts(config.afterDeploy);
    }
    catch (error) {
        console.error(error);
        return response.status(500).json({ code: 303, message: "AfterDeploy Error" }).send();
    }

    return response.json({ code: 0, message: "Successful", from: previousVersion, to: body.tag, }).send();
};
